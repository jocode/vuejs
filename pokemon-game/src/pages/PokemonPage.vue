<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>¿Quién es este pokemon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </template>
  </div>
</template>

<script>
import PokemonPicture from "@/components/PokemonPicture.vue";
import PokemonOptions from "@/components/PokemonOptions.vue";

import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  components: {
    PokemonPicture,
    PokemonOptions,
  },

  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
    };
  },

  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();

      const rdnInt = Math.floor(Math.random() * this.pokemonArr.length);
      this.pokemon = this.pokemonArr[rdnInt];
    },

    checkAnswer(pokemonId) {
      this.showPokemon = true;
      this.showAnswer = true;

      if (pokemonId === this.pokemon.id) {
        this.message = `¡Correcto! ${this.pokemon.name}`;
      } else {
        this.message = `¡Opss!. El pokemon era ${this.pokemon.name}`;
      }
    },

    newGame() {
      this.pokemon = null
      this.showPokemon = false
      this.showAnswer = false
      this.mixPokemonArray()
    },
  },

  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style>
</style>