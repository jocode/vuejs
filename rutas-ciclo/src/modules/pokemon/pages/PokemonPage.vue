<template>
    <h1>Pokemon Page <span>#{{ pokemonId }}</span></h1>

    <div v-if="pokemon">
        <img :src="pokemon.sprites.front_default" :alt="pokemon.name">
    </div>
</template>

<script>
export default {

    data(){
        return {
            pokemon: null
        }
    },

    props: {
        pokemonId: {
            type: Number,
            required: true
        }
    },

    created() {
        // const { pokemonId } = this.$route.params;
        // this.pokemonId = pokemonId
        // console.log(this.$route);
        this.getPokemon()
    },

    methods: {
        async getPokemon() {
            try {
                const pokemon = await fetch(`https://pokeapi.co/api/v2/pokemon/${this.pokemonId}`)
                .then(response => response.json())
                console.log(pokemon)
                this.pokemon = pokemon
            } catch(error) {
                console.log(error)
                this.$router.push('/')
            }
        }
    },

    watch: {
        pokemonId() {
            this.getPokemon()
        }
    }

}
</script>
