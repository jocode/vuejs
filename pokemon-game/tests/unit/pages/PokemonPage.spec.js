import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage.vue'
import { pokemons } from '../mocks/pokemon.mock'

describe('PokemonPage component', () => {

    let wrapper 

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })

    test('debe de hacer match con el match del snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de llamar a micPokemonArray al montar', () => {
        const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        wrapper = shallowMount(PokemonPage)
        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('debe de hacer match con el snapshot cuando se cargan los pokemones', () => {
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    test('debe de mostrar los componentes de pokemonPicture y PokemonsOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })

        const pictureComponent = wrapper.find('pokemon-picture-stub')
        const optionsComponent = wrapper.find('pokemon-options-stub')

        expect(pictureComponent.exists()).toBe(true)
        expect(optionsComponent.exists()).toBe(true)

        // PokemonPicture attribute pokemonid === 5
        expect(pictureComponent.attributes('pokemonid')).toBe('5')

        // PokemonsOptions attribute pokemons  toBe true
        expect(optionsComponent.attributes('pokemons')).toBeTruthy()

    })

    test('pruebas con "checkAnswer"', async() => {
        const wrapper = shallowMount(PokemonPage, {
            data(){
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: "",
                }
            }
        })

        await wrapper.vm.checkAnswer(5)

        expect(wrapper.find('h2').text()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`¡Correcto! ${pokemons[0].name}`)

        // Check when answer is wrong
        await wrapper.vm.checkAnswer(4)
        expect(wrapper.vm.showAnswer).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`¡Opss!. El pokemon era ${pokemons[0].name}`)
        expect( wrapper.vm.message ).toBe(`¡Opss!. El pokemon era ${pokemons[0].name}`)
    })

})