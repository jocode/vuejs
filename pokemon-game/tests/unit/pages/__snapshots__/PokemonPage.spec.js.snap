// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe de hacer match con el match del snapshot 1`] = `<h1>Espere por favor...</h1>`;

exports[`PokemonPage component debe de hacer match con el snapshot cuando se cargan los pokemones 1`] = `
<div>
  <h1>¿Quién es este pokemon?</h1>
  <pokemon-picture-stub pokemonid="5" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
